pipeline/util/urllib.py

```diff
@@ -36,39 +39,47 @@ def download(
-
-        headers = {'User-Agent': user_agent}
-        if not force and os.path.exists(finalfilename):
-            pre_existing_file_last_modified = datetime.utcfromtimestamp(os.path.getmtime(finalfilename))
-            headers['If-Modified-Since'] = pre_existing_file_last_modified.strftime('%a, %d %b %Y %H:%M:%S GMT')
-
-        response = get(source, timeout=timeout, headers=headers, stream=True)
-        if response.status_code == 304:
-            # File not modified since last download
-            return
-
-        with open(tmpfilename, 'wb') as file:
-            for chunk in response.iter_content(chunk_size=1024):
-                if chunk:
-                    file.write(chunk)
-
-        last_modified = None
-        last_modified_str = response.headers.get('Last-Modified')
-        if last_modified_str:
-            last_modified = parsedate_to_datetime(last_modified_str)
-            os.utime(tmpfilename, (last_modified.timestamp(), last_modified.timestamp()))
-
-        if gzip:
-            with open(tmpfilename, 'rb') as f_in, gzip.open(tmpfilename + '.gz', 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
-                if last_modified:
-                    os.utime(tmpfilename + '.gz', (last_modified.timestamp(), last_modified.timestamp()))
-                os.replace(tmpfilename + '.gz', finalfilename + '.gz')
-
-        os.replace(tmpfilename, finalfilename)
+    final_filename = Path(destination_path) / filename
+    tmp_filename = final_filename.parent / (final_filename.name + '.tmp')
+
+    if not final_filename.parent.exists():
+        final_filename.parent.mkdir(parents=True)
+
+    headers = {'User-Agent': user_agent}
+    if not force and final_filename.exists():
+        pre_existing_file_last_modified = datetime.utcfromtimestamp(final_filename.stat().st_mtime)
+        headers['If-Modified-Since'] = pre_existing_file_last_modified.strftime('%a, %d %b %Y %H:%M:%S GMT')
+
+    response = get(
+        source,
+        timeout=timeout,
+        headers=headers,
+        stream=True,
+        auth=auth,
+    )
+    if response.status_code == 304:
+        # File not modified since last download
+        return
+
+    with tmp_filename.open('wb') as file:
+        for chunk in response.iter_content(chunk_size=1024 * 1024):
+            if chunk:
+                file.write(chunk)
+
+    last_modified = None
+    last_modified_str = response.headers.get('Last-Modified')
+    if last_modified_str:
+        last_modified = parsedate_to_datetime(last_modified_str)
+        os.utime(tmp_filename, (last_modified.timestamp(), last_modified.timestamp()))
+
+    if gzip:
+        tmp_gzip_filename = tmp_filename.parent / (tmp_filename.name + '.gz')
+        final_gzip_filename = final_filename.parent / (final_filename.name + '.gz')
+
+        with tmp_filename.open('rb') as f_in, gzip.open(tmp_gzip_filename, 'wb') as f_out:
+            shutil.copyfileobj(f_in, f_out)
```

Comment:
This operation is not atomic right? Since you decided to put the tmp file right next to the final file, why not use `mv`?

Reply:
Not sure, if I understand correctly the issue you see here. `shutil.copyfileobj` zips the tmp file into a temporary gzip. Both of them are moved (via `replace`) afterwards.